### PR TITLE
[RFC] vim-patch:7.4.1640,7.4.1647,7.4.1650,7.4.1664

### PR DESCRIPTION
**Pull request by jamessan**
#### vim-patch:7.4.1640

Problem:    Crash when an autocommand changes a quickfix list. (Dominique)
Solution:   Check wether an entry is still valid. (Yegappan Lakshmanan,
            Hirohito Higashi)

https://github.com/vim/vim/commit/ffec3c53496d49668669deabc0724ec78e2274fd

#### Bump all nvim-specific error codes above E5000

In order to not conflict with new error codes that Vim adds, all Neovim
error codes should be above 5000.  The three existing sub-5000 error
codes (E926, E951, and E952) are now E50003, E5004, and E5005
respectively.

E953 was removed in 6167ce6df2753d5474ad49aea19f5957128ab015, so just
remove it from the help.

#### vim-patch:7.4.1647

Problem:    Using freed memory after setqflist() and ":caddbuffer".  (Dominique)
Solution:   Set qf_ptr when adding the first item to the quickfix list.

https://github.com/vim/vim/commit/8b20179c657b4266dff115486ca68c6a50324071

#### vim-patch:7.4.1650

Problem:    Quickfix test fails.
Solution:   Accept any number of matches.

https://github.com/vim/vim/commit/f68f1d70799631d38461c36cd59d08cf839b010d

#### vim-patch:7.4.1664

Problem:    Crash in :cgetexpr.
Solution:   Check for NULL pointer. (Dominique) Add a test.

https://github.com/vim/vim/commit/89c64d557dbe0bacfdd7b2872411b00cc1523d85

#### Mark 7.4.1697 as applied

The patch was merged in 9e1c6596 but version.c didn't get updated
accordingly.

### CODE REVIEW

File: test/functional/legacy/quickfix_spec.lua [hunk 1]
```diff
@@ -478,6 +528,23 @@ describe('helpgrep', function()
...
+  end)
+
+  it('does not crash after using caddbuffer with an empty qf list', function()
+    call('Test_caddbuffer_to_empty')
+    expected_empty()
+  end)
+
+  it('cgetexpr does not crash with a NULL element in a list', function()
+    execute('cgetexpr [$x]')
+    eq(2, eval('1+1'))
```
Comment (justinmk) [inline test/functional/legacy/quickfix_spec.lua]: I've been tagging these cases with a `-- Still alive?` comment with the intention of extracting it to a e.g. `helpers.is_alive()` function.

**Comment by jamessan** [comment]
I got as far as I could with quickfix related patches until I ran into things that require #5529 to be merged. :)

File: test/functional/legacy/quickfix_spec.lua (hunk 1 above)
Comment (jamessan) [inline test/functional/legacy/quickfix_spec.lua]: Noted.  Committed locally.  Will update once this CI round is done.

**Comment by Shougo** [comment]
The next quickfix patches are (below 2000):

1752, 1768, 1802, 1813, 1815, 1841, 1871, 1881, 1882, 1884, 1912,
1937, 1941,1949, 1950, 1953, 1964, 1966, 1984

**Pull request merged**